### ISSUE TITLE
Classify and reorganize the metric collection methods

### ISSUE THREAD

**Issue by yunseong** [snuspl/cruise#45]
This issue is related to #26, #42.

Currently We have multiple ways to collect metrics: EM has `HTrace` and Dolphin has `MetricTrackers`. Moreover, there was a suggestion of using another interface such as `Counter` in Tez. At this point, we may have to minimize the variation, and clarify which methods will be used.

First, I'd like to divide into two categories by purpose:

1) Just to investigate the apps behavior

We decided to use the HTrace in EM for the performance profiling. It seems that HTrace works very well, and I think we can apply HTrace in the same way to other systems (especially REEF) as well. 

After @dafrista spent a painful time, it turned out very difficult to abstract the trace code. As a result, the trace part should work separately.

2) To utilize the metrics

Not just watching the metrics, Dolphin actually uses them for the runtime optimization. After offline discussion, we agreed to have a separate mechanism from 1.

In this case, many components would want to share the service (e.g., Dolphin, Shuffle, etc). I think we can achieve it by generalizing and extending the current `MetricTrackerService` (as we did in REEF's `NetworkService`). I'll discuss with @gwsshs22 how Shuffle wants to use it.

In summary, 
- It seems we should have two different categories to collect metrics for: 1) Trace, and 2) Reconfiguration. 
- It would be desirable to unify the use cases (in 2) with `MetricTrackerService`.

I think this issue can be closed once we have a common service for the metric collection.

Any comments or questions are welcomed :)
